htlcswitch/switch.go

```diff
@@ -300,11 +309,39 @@ func (s *Switch) handleLocalDispatch(payment *pendingPayment, packet *htlcPacket
 	case *lnwire.UpdateFailHTLC:
 		// Retrieving the fail code from byte representation of error.
 		var userErr error
-		if code, err := htlc.Reason.ToFailCode(); err != nil {
-			userErr = errors.Errorf("can't decode fail code id"+
-				"(%v): %v", htlc.ID, err)
+
+		failure, err := payment.deobfuscator.Deobfuscate(htlc.Reason)
+		if err != nil {
+			userErr = errors.Errorf("unable to de-obfuscate "+
+				"onion failure, htlc with id(%v): %v", htlc.ID, err)
+			log.Error(userErr)
 		} else {
-			userErr = errors.New(code)
+			// Process payment failure by updating the lightning network
+			// topology by using router subsystem handler.
+			var update *lnwire.ChannelUpdate
+
+			switch failure := failure.(type) {
+			case *lnwire.FailTemporaryChannelFailure:
+				update = failure.Update
+			case *lnwire.FailAmountBelowMinimum:
+				update = &failure.Update
+			case *lnwire.FailFeeInsufficient:
+				update = &failure.Update
+			case *lnwire.FailIncorrectCltvExpiry:
+				update = &failure.Update
+			case *lnwire.FailExpiryTooSoon:
+				update = &failure.Update
+			case *lnwire.FailChannelDisabled:
+				update = &failure.Update
+			}
+
+			if update != nil {
+				log.Info("Received payment failure(%v), applying lightning "+
+					"network topology update", failure.Code())
+				s.cfg.UpdateTopology(update)
```

Comment:
I would prefer to do that in another `PR` just because when I started to changing the way how we return the error, I realised that we a mixing local and remote errors, for that reason we can't just swap the `error` to the `lnwire.FailureMessage`, and it would require a bit of the thinking. Would like to do that after `PR` will be merged.